pkg/util/strategicpatch/patch.go

```diff
@@ -1447,3 +1453,34 @@ func GetServerSupportedSMPatchVersion(discoveryClient discovery.DiscoveryInterfa
 	}
 	return Unknown, fmt.Errorf("The version is too old: %v\n", serverVersion)
 }
+
+// GuidedJsonMerge merges two JSON formatted items by following the structure of dataStruct's type. We use
+// dataStruct's type to respect strategicMerge guidelines and ensure the structure of the JSON items
+// (e.g. no fields are missing, no additional fields exist, etc)
+func GuidedJsonMerge(j1, j2 []byte, dataStruct interface{}) ([]byte, error) {
+	j1Map := map[string]interface{}{}
+	if len(j1) > 0 {
+		if err := json.Unmarshal(j1, &j1Map); err != nil {
+			return nil, fmt.Errorf("Invalid JSON document")
+		}
+	}
+
+	j2Map := map[string]interface{}{}
+	if len(j2) > 0 {
+		if err := json.Unmarshal(j2, &j2Map); err != nil {
+			return nil, fmt.Errorf("Invalid JSON document")
+		}
+	}
+
+	t, err := getTagStructType(dataStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	mergedMap, err := mergeMap(j1Map, j2Map, t, false)
```

Comment:
This function is a general "frontend" function to merge 2 json with respect to the `dataStruct` type guidelines (e.g. which fields have strategicMerge attributes). In our case though we use this generic merge under the assumption that j1 and j2 are the same object except 3 differences:

1. j2 will have fields that will have the value `null` that j1 will not have (so when we merge, those fields together with the identical fields of j1 and j2 will be simply added to the end result)
2. j2 null-valued fields might have to overwrite empty structs on j1 (see the example below)
3. extra values added into j1 (e.g. annotations from apply) are again just forwarded to the end result

#### Example on 2

User gives the following Deployment configuration:
```
apiVersion: extensions/v1beta1
kind: Deployment
...
    strategy: null
...
```

When `kubectl` Unmarshals this object into a Go struct, the `strategy` field will not be `nil` but will be an empty `DeploymentStrategy` struct. This is due to [this](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/types.go#L192) (It is NOT a pointer, it is a struct). So when we have to merge j1 and j2, j2 will have `strategy: null` and j1 will have `strategy: {}`. This is a conflict that we resolve in favor of j2 (ie we overwrite `{}` with `null`)


I tried to give an explanation about this inside `apply.go` at the point where i call `GuidedJsonMerge`, if you thing this is not enough or doesn't make sense please tell me so I can add some more details from this post here.

## IMPORTANT
BTW, I have concern about: 3. If a user's config is changed dramatically during loading (e.g. A user gives a Secret and during Encoding we delete `stringData` and we transform them into `data`) then my simplistic approach to just merge the source object with the modified `versionedObject` will fail. From my understanding any kind of object transformation is happening only at the API server and only during Decoding?